logsearchapi/server/db-migrations.go

```diff
@@ -0,0 +1,150 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2022 MinIO, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package server
+
+import (
+	"context"
+	"log"
+
+	"github.com/lib/pq"
+)
+
+// dbMigration represents a DB migration using db-client c.
+// Note: a migration func should be idempotent.
+type dbMigration func(ctx context.Context, c *DBClient) error
+
+var allMigrations = []dbMigration{
+	addAccessKeyColAndIndex,
+	addAuditLogIndices,
+	addReqInfoIndices,
+
+	// Add new migrations here below
+}
+
+func (c *DBClient) runMigrations(ctx context.Context) error {
+	for _, migration := range allMigrations {
+		if err := migration(ctx, c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func duplicateColErr(err error) bool {
+	if pqerr, ok := err.(*pq.Error); ok &&
+		pqerr.Code == "42701" {
+		return true
+	}
+	return false
+}
+
+func duplicateTblErr(err error) bool {
+	if pqerr, ok := err.(*pq.Error); ok &&
+		pqerr.Code == "42P07" {
+		return true
+	}
+	return false
+}
+
+func (c *DBClient) runQueries(ctx context.Context, queries []string, ignoreErr func(error) bool) error {
+	for _, query := range queries {
+		if _, err := c.ExecContext(ctx, query); err != nil {
+			if ignoreErr(err) {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func updateAccessKeyCol(ctx context.Context, c *DBClient) {
+	updQ := `WITH req AS (
+                             SELECT log->>'requestID' AS request_id,
+                                    COALESCE(
+                                       substring(
+                                           log->'requestHeader'->>'Authorization',
+                                           e'^AWS4-HMAC-SHA256\\s+Credential\\s*=\\s*([^/]+)'
+                                       ),
+                                       substring(log->'requestHeader'->>'Authorization', e'^AWS\\s+([^:]+)')
+                                    ) AS access_key
+                               FROM audit_log_events AS a JOIN request_info AS b ON (a.event_time = b.time)
+                              WHERE b.access_key IS NULL
+                           ORDER BY event_time
+                              LIMIT $1
+                             OFFSET $2
```

Comment:
One other improvement we can do here is that we can omit the offset and simply repeat the query until the number of rows affected become zero.

Reply:
Otherwise it's actually a bug.